### DISCUSSION TITLE
How does `collect` (with ordinary streams) differ from `do`?

### DISCUSSION THREAD

**Discussion by webbedspace** [nushell/nushell#6894] [Q&A] [answered]
I feel like these work identically (in the sense that any `collect { ... }` without --keep-env can be rewritten as `do { ... } $in`.) However, the phrase "collect the stream" gives me pause. Does `collect` do something different to raw streams compared to `do`? Or are they really the same?

Note: I do NOT mean the lazyframe `collect`.

**Comment by theAkito** [comment]
I already see one difference in your post.

```nu
collect { ... }
```
versus

```nu
do { ... } $in
```

The first one is shorter, in terms of word-like blocks.

**Reply by webbedspace** [reply]
It's not actually shorter in terms of character count, though, which I think matters more. And also, I consider this a fairly meaningless "difference".

**Reply by theAkito** [reply]
Character count matters tons, if any special character is involved. For example, `str::replace` takes up an eternity to type compared to `str replace`, because space is easiest & fastest to type for everyone with a normal keyboard, while colons are comparatively tougher; even more so on some non-English layouts.

In this case, we have a dollar sign, which is tons harder to reach than just writing `collect` rather than `do`.

>I consider this a fairly meaningless "difference".

As I explained above, it may be a huge difference. You are mainly talking out of your own view, but I do not see how you are trying to see a general perspective from other users.

For example, to me personally, it seems a million times easier to just type a single word and issue a block, rather than doing that plus an additional "meaningless" variable after it.
If you are very strict about it, then the appended variable is essentially just bloat and if that wasn't bad enough, it's even way harder to type, because it contains a special character. Plus, it needs to be appended after a block, which is not great, because you have to think in cat1, cat2, cat(1|3), instead of cat1, cat2, when you type, as `collect { ... }` has only two categories of semantics & you have to only use each once, while `do { ... } $in` has two or, depending on how one would define it, three different categories of semantics. In my mind, it is three categories, as the last word is prepended with a special character.
Therefore, it is more difficult to think in the latter way, rather than the former.

____

My point is, that you see everything from your developer perspective, instead of trying to see it from a general perspective, involving the average users. Additionally, your statements make it seem a bit like you do not value UX.

**Reply by rgwood** [reply]
> Additionally, your statements make it seem a bit like you do not value UX.

@theAkito: We try to assume the best of other contributors here, and if you can't do that you will be removed from the Nushell repo. You have been warned.

**Reply by webbedspace** [reply]
@theAkito Well, first of all I'm not a developer (I only have 3 pull requests accepted as of today) and secondly, you don't actually have to write the `$in` after the `do` block: you can just write `do { ... }`, and it gets passed in automatically. The reason I write it with the trailing `$in` in this example is because I assumed using explicit `$in` performs the "collection" that `collect` does (gathering the entire stream result into a variable) which might change the semantics of the `do` block. So, the only time you'd need to write `$in` is if that collection (if it actually happens) is strictly necessary (which I believe is very infrequent, given that I personally never use `collect`).

But, like, is **anyone** here going to answer my original question:

>Does `collect` do something different to raw streams compared to `do`?

**Reply by theAkito** [reply]
> you don't actually have to write the `$in` after the `do` block: you can just write `do { ... }`, and it gets passed in automatically. The reason I write it with the trailing `$in` in this example is because I assumed using explicit `$in` performs the "collection" that `collect` does (gathering the entire stream result into a variable) which might change the semantics of the `do` block. So, the only time you'd need to write `$in` is if that collection (if it actually happens) is strictly necessary (which I believe is very infrequent, given that I personally never use `collect`).

Okay, now I understand. Thanks for the thorough clarification.

Is it possible for you to test the difference between the two manually in scripts?
I did not have any experience with collections of this type yet, so I would need to go into it first.

> first of all I'm not a developer (I only have 3 pull requests accepted as of today)

Don't be coy, you are a developer. 😄 The way you phrased the initial question, plus the topic of the matter in the first place is something a developer would think about. A non-developer most likely wouldn't even think about it.

**Comment by rgwood** [comment, marked as answer]
I'm not able to answer your question 100% definitively; it's a little hard to follow the code and I ran out of time+steam. The `collect` code explicitly converts input to a `Value` and I _think_ `do` is implicitly doing the same thing.

And so I suspect we don't need both `collect` and `do`, at least in their current forms. I brought that up on Discord and JT agrees.
